libbeat/processors/add_kubernetes_metadata/podwatcher.go

```diff
@@ -189,18 +194,65 @@ func (p *PodWatcher) worker() {
 	}
 }
 
+// Check annotations flagged as deleted for their last access time, fully delete
+// the ones older than p.cleanupTimeout
+func (p *PodWatcher) cleanupWorker() {
+	for {
+		// Wait a full period
+		time.Sleep(p.cleanupTimeout)
+
+		select {
+		case <-p.ctx.Done():
+			return
+		default:
+			// Check entries for timeout
+			var toDelete []string
+			timeout := time.Now().Add(-p.cleanupTimeout)
+			p.annotationCache.RLock()
+			for key, lastSeen := range p.annotationCache.deleted {
+				if lastSeen.Before(timeout) {
+					toDelete = append(toDelete, key)
+				}
+			}
+			p.annotationCache.RUnlock()
+
+			// Delete timed out entries:
+			p.annotationCache.Lock()
+			for _, key := range toDelete {
+				delete(p.annotationCache.deleted, key)
+				delete(p.annotationCache.annotations, key)
+			}
+			p.annotationCache.Unlock()
+		}
+	}
+}
+
 func (p *PodWatcher) GetMetaData(arg string) common.MapStr {
-	p.annotationCache.Lock()
-	defer p.annotationCache.Unlock()
-	if meta, ok := p.annotationCache.annotations[arg]; ok {
+	p.annotationCache.RLock()
+	meta, ok := p.annotationCache.annotations[arg]
+	var deleted bool
+	if ok {
```

Comment:
For logs, we can update the pod labels using `kubectl label po`. The container name remains the same and the update will call `onPodDelete && onPodAdd`. So now the living container's metadata is in the deleted map. This portion of code will end up returning old metadata.

Reply:
Ideally we should first check metadata from live pods and return data for that.

Reply:
I've added a commit addressing this, basically, if an add method is called it will remove any deleted flag for all its indexes